File: 2025-10/2025-10-18/missing_socks.py
def sock_pairs(pairs, cycles):
    remaining_pairs = (pairs * 2 
        - cycles // 2 
        + cycles // 3 
        - cycles // 5 
        + cycles // 10 * 2) \
        // 2

    return remaining_pairs if remaining_pairs >= 0 else 0

print(sock_pairs(2, 5))
print(sock_pairs(1, 2))
print(sock_pairs(5, 11))
print(sock_pairs(6, 25))
print(sock_pairs(1, 8))

# --- TEST SUITE ---

tests_text = '''
1. sock_pairs(2, 5) should return 1.
2. sock_pairs(1, 2) should return 0.
3. sock_pairs(5, 11) should return 4.
4. sock_pairs(6, 25) should return 3.
5. sock_pairs(1, 8) should return 0.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)

File: 10 October/10-18/missingSocks.js
function sockPairs(pairs, cycles) {
  const remainingPairs = Math.floor(
    (pairs * 2 
    - Math.floor(cycles / 2) 
    + Math.floor(cycles / 3) 
    - Math.floor(cycles / 5) 
    + Math.floor(cycles / 10) * 2) 
    / 2);

  return remainingPairs >= 0 ? remainingPairs : 0;
}

console.log(sockPairs(2, 5));
console.log(sockPairs(1, 2));
console.log(sockPairs(5, 11));
console.log(sockPairs(6, 25));
console.log(sockPairs(1, 8));

// --- TEST SUITE ---

const testsText = `
1. sockPairs(2, 5) should return 1.
2. sockPairs(1, 2) should return 0.
3. sockPairs(5, 11) should return 4.
4. sockPairs(6, 25) should return 3.
5. sockPairs(1, 8) should return 0.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (const index in a) {
    if (Array.isArray(a[index])) {
      if (Array.isArray(b[index])) {
        if (!arraysEqual(a[index], b[index]))
          return false;
      }
      else 
        return false;
    }

    else if (a[index] !== b[index])
      return false;
  }

  return true;
}

runTests(testData);